File: android/SwiftKotlination/app/src/main/kotlin/fr/jhandguy/swiftkotlination/features/story/viewModel/StoryViewModel.kt
package fr.jhandguy.swiftkotlination.features.story.viewModel

import android.graphics.Bitmap
import fr.jhandguy.swiftkotlination.features.story.factory.StoryFactory
import fr.jhandguy.swiftkotlination.features.story.model.Story
import fr.jhandguy.swiftkotlination.observer.Disposable
import fr.jhandguy.swiftkotlination.observer.Observer
import fr.jhandguy.swiftkotlination.observer.Result

class StoryViewModel(factory: StoryFactory, story: Story = Story(), var image: Bitmap? = null) {
    private val storyManager = factory.makeStoryManager(story)
    private val imageManager = factory.makeImageManager()

    suspend fun story(observer: Observer<Story>) = storyManager.story(observer)

    suspend fun image(url: String, observer: Observer<Bitmap>): Disposable? = image?.let { image ->
        observer(Result.Success(image))
        return null
    } ?: imageManager.image(url) { result ->
        when (result) {
            is Result.Success -> {
                image = result.data
                observer(result)
            }
            is Result.Failure -> observer(result)
        }
    }
}

File: ios/StoryKit/StoryKit/Sources/ViewModel/StoryViewModel.swift
import ImageKit
import NetworkKit
import UIKit.UIImage

public final class StoryViewModel {
    public typealias Factory = ImageFactory & StoryFactory
    private let factory: Factory
    private let story: Story

    private lazy var storyManager = factory.makeStoryManager(for: story)
    private lazy var imageManager = factory.makeImageManager()

    private var image: UIImage?

    public init(factory: Factory, story: Story) {
        self.factory = factory
        self.story = story
    }
}

extension StoryViewModel {
    func story(_ observer: @escaping Observer<Story>) {
        storyManager.story(observer)
    }

    @discardableResult
    func image(with url: String, _ observer: @escaping Observer<UIImage>) -> Disposable? {
        guard let image = image else {
            return imageManager
                .image(with: url) { [weak self] result in
                    switch result {
                    case let .success(data):
                        guard let image = UIImage(data: data) else {
                            observer(.failure(NetworkError.invalidData))
                            return
                        }
                        self?.image = image
                        observer(.success(image))

                    case let .failure(error):
                        observer(.failure(error))
                    }
                }
        }

        observer(.success(image))
        return nil
    }
}
